### DISCUSSION TITLE
How to

### DISCUSSION THREAD

**Discussion by JDRanpariya** [sqlalchemy/sqlalchemy#5892] [Usage Questions] [answered]
**Describe your question**
I'm having table A with foreign key of table B. Now When I'm adding a New object to table A with following code.
it gives me the following error describe below.

**Example**
[Proof 1 see 2nd answer it works fine there](https://stackoverflow.com/questions/33083772/sqlalchemy-attributeerror-str-object-has-no-attribute-sa-instance-state)

**Complete stack trace**
```
problem occured on line session.commit()
sqlalchemy.exc.IntegrityError: (psycopg2.errors.UniqueViolation) duplicate key value violates unique constraint "ratings_rating_key"
DETAIL:  Key (rating)=(8.8) already exists.

[SQL: INSERT INTO ratings (rating) VALUES (%(rating)s) RETURNING ratings.id]
[parameters: {'rating': 8.8}]
```
**Versions**
 - OS: Windows10
 - Python: 3.8.3
 - SQLAlchemy: 1.3.22
 - Database: postgres
 - DBAPI: 

**Additional context**
Code
```pipeline.py
session = self.Session()
        performer = Performer(
                    name = item['name'],
                    aliases = item['aliases'],
                    gender = item['gender'],
                    description = item['description'],
                    profile_pic = item['profile_pic'],
                    date_of_birth = item['date_of_birth'], # get age from here
                    ethnicity = item['ethnicity'],
                    birth_place = item['birth_place'],
                    height = item['height'],
                    
        )

        rating = session.query(Rating).filter_by(rating=item['rating']).first()
        rating_exists = rating is not None

        if rating_exists:
            performer.rating = rating
        else:
            performer.rating = Rating(rating=item['rating'])

        performer_exists = session.query(Performer).filter_by(name=item['name']).first() is not None

        if performer_exists:
            logging.info(f'Item {performer} is in db')
            return item
        else:
            try:
                session.add(performer)
                session.commit()
                logging.info(f'Item {performer} stored in db')
            except:
                logging.info(f'Failed to add {performer} to db')
                session.rollback()
                raise
            finally:
                session.close()
```

**Have a nice day!**
The really thing is I don't know how to add performer and add it but when when rating exist it should use that or other wise insert so basically how to add some fields of model and update some specific ones.

**Comment by zzzeek** [comment]
> **Example**
> [Proof 1 see 2nd answer it works fine there](https://stackoverflow.com/questions/33083772/sqlalchemy-attributeerror-str-object-has-no-attribute-sa-instance-state)

I don't see anything in the linked stackoverflow answer that appears to be relevant here.


> 
> **Complete stack trace**
> 
> ```
> problem occured on line session.commit()
> sqlalchemy.exc.IntegrityError: (psycopg2.errors.UniqueViolation) duplicate key value violates unique constraint "ratings_rating_key"
> DETAIL:  Key (rating)=(8.8) already exists.
> 
> [SQL: INSERT INTO ratings (rating) VALUES (%(rating)s) RETURNING ratings.id]
> [parameters: {'rating': 8.8}]

this is a unique constraint violation.  it means that the table you are inserting into has a [UNIQUE CONSTRAINT](https://www.postgresqltutorial.com/postgresql-unique-constraint/) on it which prevents more than one row from having a particular value.

As to what that value is, I can't answer because I would need to see the definition of this unique constraint and the table schema it refers towards as well as the SQLAlchemy model, the name of the constraint itself does not tell me what the actual column(s) in question are.   I guess it's "rating", so if this is a decimal or floating point value perhaps there is a floating point conversion issue in play, such as your Python floating point "rating" has more precision than what's in the database and things aren't matching up.   just a guess, I've never seen that before but that's the only odd thing I can see here.

you want to go to your actual program, set `echo="debug"` on your create_engine() call, then walk through your program step by step to see what rows are returned for what queries.

**Reply by JDRanpariya** [reply]
value coming from python
```
2021-01-30 14:04:37 [root] INFO: 8.8
2021-01-30 14:04:37 [root] INFO: <class 'float'>
```
value field for database
```models.py
class Performer(DeclarativeBase):
    """ c """
    __tablename__ = 'performers'

    id = Column(Integer, primary_key=True)
    name = Column('name', String)
    aliases = Column('aliases', Text(), nullable=True)
    description = Column('description', Text(), nullable=True)
    profile_pic = Column('profile_pic', ARRAY(String))
    date_of_birth = Column('date_of_birth', Date, nullable=True) # get age from here

    #repetative fields
    rating_id = Column(Integer, ForeignKey('ratings.id'))
    rating = relationship("Rating", back_populates='performers')


class Rating(DeclarativeBase):
    """ c """
    __tablename__ = 'ratings'

    id = Column(Integer, primary_key=True)
    rating = Column('rating', Float(precision=1), unique=True)

    performers =relationship("Performer", back_populates='rating')

    scenes = relationship("Scene", back_populates='rating')

    movies = relationship("Movie", back_populates='rating')
```

**Comment by JDRanpariya** [comment]
Btw This is my first time using relationships and databases. So If you find anything wired or not appropriate Please suggest the changes.
The relationship between Performer & Rating is Many to One respectively. As There can be lot of performers with rating of 9 and so on.

**Comment by zzzeek** [comment, marked as answer]
https://www.postgresql.org/docs/10/datatype-numeric.html#DATATYPE-FLOAT

> Comparing two floating-point values for equality might not always work as expected.

the precision is ignored with that datatype for Postgresql.  Try using Numeric() instead.

**Reply by JDRanpariya** [reply]
Hi that really solved the problem. I wonder how you guessed it right. Btw Now I'm having another problem and that's like Whenever adding a new performer I check using `session.query(Performer).filter_by(name=item['name']).first()` if performer exists but it showing True all the time except when Rating is different.
```pipelines.py
def process_item(self, item, spider):

        session = self.Session()
        performer = Performer(
                    name = item['name'],
                    aliases = item['aliases'],
                    gender = item['gender'],
                    description = item['description'],
                    profile_pic = item['profile_pic'],
                    date_of_birth = item['date_of_birth'], # get age from here                    
        )

        rating = session.query(Rating).filter_by(rating=item['rating']).first()
        
        if rating is not None:
            performer.rating = rating
        else:
            performer.rating = Rating(rating=item['rating'])

        performer_exists = session.query(Performer).filter_by(name=item['name']).first() is not None

        if performer_exists:
            logging.info(item['name'])
            logging.info(session.query(Performer).filter_by(name=item['name']).first())
            logging.info(f'Item {performer} is in db')

            return item
        else:
            try:
                session.add(performer)
                session.commit()
                logging.info(f'Item {performer} stored in db')
            except:
                logging.info(f'Failed to add {performer} to db')
                session.rollback()
                raise
            finally:
                session.close()
        return item
```
Traceback looks like ( it's saying Performer is in database even it's new I'm wondering How ? )
```
2021-01-31 11:54:43 [root] INFO: new_performer
2021-01-31 11:54:43 [root] INFO: new_performer
2021-01-31 11:54:43 [root] INFO: Item <abc_scraper.models.Performer object at 0x05FA2CE8> is in db
```

I'm checking before adding and commiting

**Comment by JDRanpariya** [comment]
looks like my rating_id column is acting like primerykey, here is what my table looks like
| id | all other cols |raing_id |
| :---: | :---: | :---: |
| 1 | -- | 1 |
|2|--|2|
|19|--|3|

somehow performers with same rating_id isn't showing up

so, Rating here is parent ( one ) and performer is child ( many ) and problem is child is not getting added because parent exists.

Thank you so much, You were very helpful.

**Reply by zzzeek** [reply]
when you do this:

    performer.rating = rating

that Performer is getting placed into the Session.    background on this behavior is at https://docs.sqlalchemy.org/en/13/orm/cascades.html#controlling-cascade-on-backrefs and I recommend setting this.   the behavior is also deprecated: https://docs.sqlalchemy.org/en/14/changelog/migration_14.html#cascade-backrefs-behavior-deprecated-for-removal-in-2-0

then you emit a query(), it's inserting your Performer object right there due to autoflush.   so performer_exists will always be True.

**Reply by JDRanpariya** [reply]
I'am really thankful I checked docs and used `future=True` and `backref_cascade=False` When using` performer.rating = rating` and It worked all fine. I'm wanna check two answers( to green) but thats not possible.
